cpu/nrf51/radio/nrf51_ble/nrf51_ble.c

```diff
@@ -0,0 +1,628 @@
+
+    NRF_RADIO->PCNF1 = (CONF_WHITENING << RADIO_PCNF1_WHITEEN_Pos) |
+                       (CONF_ENDIAN << RADIO_PCNF1_ENDIAN_Pos) |
+                       (CONF_BASE_ADDR_LEN << RADIO_PCNF1_BALEN_Pos) |
+                       //(CONF_STATLEN << RADIO_PCNF1_STATLEN_Pos) |
+                       (CONF_PAYLOAD_LEN << RADIO_PCNF1_MAXLEN_Pos);
+    /* configure CRC unit */
+    NRF_RADIO->CRCCNF = RADIO_CRCCNF_LEN_Three << RADIO_CRCCNF_LEN_Pos |
+                        RADIO_CRCCNF_SKIPADDR_Skip << RADIO_CRCCNF_SKIPADDR_Pos;
+
+    NRF_RADIO->CRCPOLY = CONF_CRC_POLY;
+    NRF_RADIO->CRCINIT = CONF_CRC_INIT;
+    NRF_RADIO->DATAWHITEIV = BLE_DEFAULT_CHANNEL;
+
+    /* set shortcuts for more efficient transfer */
+    NRF_RADIO->SHORTS = (1 << RADIO_SHORTS_READY_START_Pos);
+    /* enable interrupts */
+    NVIC_EnableIRQ(RADIO_IRQn);
+    /* enable END interrupt */
+    NRF_RADIO->EVENTS_END = 0;
+    NRF_RADIO->INTENSET = (1 << RADIO_INTENSET_END_Pos);
+
+    /* put device in receive mode */
+    _switch_to_rx();
+
+    return 0;
+}
+
+/* BLE interface low-level receive function */
+static void _receive_data(void)
+{
+    ble_adv_pkt_t *data;
+    gnrc_pktsnip_t *pkt;
+    uint16_t pkt_size;
+
+    /* only read data if we have somewhere to send it to */
+    if (_netdev->event_cb == NULL) {
+        DEBUG("nrf51_ble: receive_data (return)\n");
+        return;
+    }
+
+    /* get pointer to RX data buffer */
+    data = &(_rx_buf[_rx_next ^ 1]);
+
+    pkt_size = BLE_ADV_HDR_LEN + data->header.length;
+
+    /* allocate and fill payload */
+    pkt = gnrc_pktbuf_add(NULL, data, pkt_size, GNRC_NETTYPE_UNDEF);
+
+    if (pkt == NULL) {
+        DEBUG("nrf51_ble: Error allocating packet payload on RX\n");
+        gnrc_pktbuf_release(pkt);
+        return;
+    }
+
+    /* pass on the received packet */
+    _netdev->event_cb(NETDEV_EVENT_RX_COMPLETE, pkt);
+}
+
+/* BLE interface low-level send function */
+int _send(gnrc_netdev_t *dev, gnrc_pktsnip_t *pkt)
+{
+    (void)dev;
+    size_t size;
+
+    /* check packet */
+    if (pkt == NULL) {
+        DEBUG("nrf51_ble: Error sending packet: packet incomplete\n");
+        return -ENOMSG;
+    }
+
+    /* check if payload is withing length bounds */
+    size = gnrc_pkt_len(pkt);
+
+    if (size > CONF_PAYLOAD_LEN) {
+        gnrc_pktbuf_release(pkt);
+        DEBUG("nrf51_ble: Error sending packet: invalid packet length\n");
+        return -EOVERFLOW;
+    }
+
+    memcpy(&(_tx_buf), pkt->data, size);
+
+    /* wait for any ongoing transmission to finish */
+    while (_state == STATE_TX);
+
+    /* save old state and switch to idle if applicable */
+    _tx_prestate = _state;
+
+    if (_tx_prestate == STATE_RX) {
+        _switch_to_idle();
+    }
+
+    /* set packet pointer to TX buffer and write destination address */
+    NRF_RADIO->PACKETPTR = (uint32_t)(&_tx_buf);
+
+    /* start transmission */
+    _state = STATE_TX;
+    NRF_RADIO->TASKS_TXEN = 1;
+
```

Comment:
Won't it be reasonable to wait with releasing the `pkt` until the transmission has finished? Or why isn't that needed?